### DISCUSSION TITLE
What is best way to revert commit?

### DISCUSSION THREAD

**Discussion by andr1972** [community/community#22346] [New to GitHub]
<p>I have all sources of project but main.cpp in one directory and main.cpp in another directory. This caused problem with Doxygen to document all files, and I moved all but main() from main.cpp to another file.</p>
<p>Next I commit and push to github.</p>
<p>But my doxygened source are in one branch, I want merge to doxygen branch from other branches and I must revert commit to merge.</p>
<p>My revert must not only make me previous source state, but also delete commit history to do correct merge.</p>

**Comment by lee-dohm** [comment, marked as answer]
<p>There are a bunch of different ways to revert a commit, depending on what the desired end result is. It sounds like you want to:</p>
<ol>
<li>Remove the last commit on the current branch from the repository</li>
<li>Revert the state of the local repository to the commit prior to the commit named in item 1</li>
<li>Update the current branch on the <code>origin</code> remote to the local state of the current branch</li>
</ol>
<p>:rotating_light: <strong>WARNING</strong> :rotating_light:</p>
<p>Executing these commands <em>will lose history</em> (though Git does its best to make even that recoverable). Use these commands with caution.</p>
<p>In order to achieve items 1 and 2, you can execute:</p>
<pre><code>git reset --hard HEAD^
</code></pre>
<p>In order to achive item 3, you can execute:</p>
<pre><code>git push --force
</code></pre>
<p>Any time you use <code>git push --force</code> though, you may cause problems for other people working on the same branch because you’re rewriting history. So use it with caution as well.</p>
<p>I hope that helps!</p>
